opennms-dao/src/main/java/org/opennms/netmgt/dao/hibernate/AbstractDaoHibernate.java

```diff
@@ -259,8 +262,31 @@ public void deleteAll(final Collection<T> entities) throws DataAccessException {
 
     @Override
     public List<T> findMatching(final org.opennms.core.criteria.Criteria criteria) {
-        final HibernateCallback<List<T>> callback = buildHibernateCallback(criteria);
-        return getHibernateTemplate().execute(callback);
+        if(criteria.isMultipleAnd()){
+            Collection<Restriction> allRestrictions = criteria.getRestrictions();
+            Collection<Restriction> copyOfAllRestriction = new ArrayList<Restriction>(allRestrictions) ;
+            List<T> allRecords = new ArrayList<>();
+            Set<T> allUniqueRecords = new LinkedHashSet<>();
+            allRestrictions.stream().filter(restriction -> restriction.getType().equals(Restriction.RestrictionType.MULTIAND))
+                    .forEach(restriction -> {
+                        Collection<Restriction> allMultiAndRestrictions = ((AllRestriction) restriction).getRestrictions();
+                        AllRestriction[] tempRestrictionArray = new AllRestriction[allMultiAndRestrictions.size()];
+                        allMultiAndRestrictions.toArray(tempRestrictionArray);
+                        copyOfAllRestriction.removeIf(restriction::equals);
+                        Arrays.stream(tempRestrictionArray).forEach(tempRestriction -> {
+                            copyOfAllRestriction.add(tempRestriction);
+                            criteria.setRestrictions(copyOfAllRestriction);
+                            final HibernateCallback<List<T>> callback = buildHibernateCallback(criteria);
+                            allUniqueRecords.addAll(getHibernateTemplate().execute(callback));
```

Comment:
Shouldn't this be intersection of all the results not the union of all results ?

Reply:
Intersection implemented